internal/store/cluster.go

```diff
@@ -23,6 +23,99 @@ func (sqlStore *SQLStore) GetCluster(id string) (*model.Cluster, error) {
 	return &cluster, nil
 }
 
+// GetUnlockedClusterPendingWork returns an unlocked cluster in a pending state.
+func (sqlStore *SQLStore) GetUnlockedClusterPendingWork() (*model.Cluster, error) {
+	var cluster model.Cluster
+	err := sqlStore.getBuilder(sqlStore.db, &cluster, sq.
+		Select("*").
+		From("Cluster").
+		Where(sq.Eq{
+			"State": []string{
+				model.ClusterStateCreationRequested,
+				model.ClusterStateUpgradeRequested,
+				model.ClusterStateDeletionRequested,
+			},
+		}).
+		Where("LockAcquiredAt = 0").
+		OrderBy("State ASC"),
+	)
+	if err == sql.ErrNoRows {
+		return nil, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to get cluster pending work")
+	}
+
+	return &cluster, nil
+}
+
+// LockCluster marks the cluster as locked for exclusive use by the caller.
+func (sqlStore *SQLStore) LockCluster(clusterID string) (bool, error) {
+	result, err := sqlStore.execBuilder(sqlStore.db, sq.
```

Comment:
Relying on database semantics here to guarantee two updates can't both write to the same row like this.